api/services/passport.js

```diff
@@ -1,85 +1,92 @@
-      EventCreator.audit(Event.labels.AUTHENTICATION, user, eventBody);
-
-      return user.update({
-        githubAccessToken: accessToken,
-        githubUserId: profile.id,
-        signedInAt: new Date(),
-      });
-    })
-    .then(() => {
-      RepositoryVerifier.verifyUserRepos(user); // verify user's site's repos
-      callback(null, user);
-    })
-    .catch((err) => {
-      logger.warn('Authentication error: ', err);
-      callback(err);
+const {
+  github: { options: githubOptions },
+  uaa: { options: uaaOptions },
+} = config.passport;
+
+async function verifyGithub(accessToken, _refreshToken, profile, callback) {
+  try {
+    const isValidUser = await GitHub.validateUser(accessToken, false);
+    if (!isValidUser) {
+      return callback(null, false);
+    }
+
+    const username = profile.username.toLowerCase();
+    // eslint-disable-next-line no-underscore-dangle
+    const { email } = profile._json;
+
+    const [user] = await User.upsert({
+      username,
+      email,
+      githubAccessToken: accessToken,
+      githubUserId: profile.id,
+      signedInAt: new Date(),
     });
-};
 
-passport.use(new GitHubStrategy(config.passport.github.options, githubVerifyCallback));
+    EventCreator.audit(Event.labels.AUTHENTICATION, user, { action: 'login' });
 
-passport.logout = (req, res) => {
-  const { user } = req;
-  req.logout();
-  if (user) {
-    const eventBody = { action: 'logout' };
-    EventCreator.audit(Event.labels.AUTHENTICATION, user, eventBody);
+    // should we wait for this?
+    RepositoryVerifier.verifyUserRepos(user);
+    return callback(null, user);
+  } catch (err) {
+    logger.warn('Authentication error: ', err);
+    return callback(err);
   }
-  req.session.destroy(() => {
-    res.redirect(config.app.hostname);
-  });
-};
+}
+
+async function verifyUAA(accessToken, _refreshToken, profile, callback) {
+  const { email } = profile;
+
+  try {
+    const user = await User.findOne({ where: { adminEmail: email } });
 
-passport.callback = (req, res) => {
-  passport.authenticate('github')(req, res, () => {
-    if (req.user) {
-      req.session.authenticated = true;
-      req.session.authenticatedAt = new Date();
-      req.session.save(() => {
-        if (req.session.authRedirectPath) {
-          res.redirect(req.session.authRedirectPath);
-        } else {
-          res.redirect('/');
-        }
-      });
-    } else {
-      req.flash('error', {
-        title: 'Unauthorized',
-        message: 'Apologies; you don\'t have access to Federalist! '
-                 + 'Please contact the Federalist team if this is in error.',
-      });
-      res.redirect('/');
+    if (!user) {
+      return callback(null, false);
     }
-  });
+
+    await user.update({
+      signedInAt: new Date(),
+    });
+
+    EventCreator.audit(Event.labels.AUTHENTICATION, user, { action: 'login' });
+
+    // should we wait for this?
+    // do i need to do this?
+    RepositoryVerifier.verifyUserRepos(user);
```

Comment:
should we wait for this to finish?

Reply:
i don't have a strong opinion on this but I lean towards not waiting. i mainly see the benefit in waiting for a user who has not logged in for quite a while